### ISSUE TITLE
Overflow to next page

### ISSUE THREAD

**Issue by danielepolencic** [evnbr/bindery#84]
This is not a bug or a feature request, but rather a request for clarification.

Consider the following content:

![image](https://user-images.githubusercontent.com/1130499/101974348-2b495b00-3c75-11eb-92e2-3ca20c28832d.png)

The content overflows.
I'd like the red block to be always placed on a new page and I'd like to remove it from the natural flow of the document, like this:

![image](https://user-images.githubusercontent.com/1130499/101975004-5a5fcc80-3c75-11eb-979b-771dab3e3081.png)

I tried using FullPageBleed, but that does work. If I use the PageBreak, this is what I obtain:

![image](https://user-images.githubusercontent.com/1130499/101975176-719eba00-3c75-11eb-80f9-8c79fcee05be.png)

Is there any way I could code a rule to reflow the content as per scenario 3?

![Frame 8](https://user-images.githubusercontent.com/1130499/101975182-7f543f80-3c75-11eb-834e-5ff644781b22.png)

**Comment by evnbr** [comment]
Yes, I believe what you're referring to is basically a "page float" as described here https://www.w3.org/TR/css-page-floats-3/ 

Right now bindery should support placing the red element on its own page, completely out of the text flow, using FullBleedPage. But it does not support floating the element onto a new page and then continuing the text flow below on that same page.

This is a super reasonable scenario, and something I hope to enable in the next major version of bindery... part of the reason i was warning you off going too deep on fixing the rules implementation in #83 :)

**Comment by danielepolencic** [comment]
Understood. Is there a rough roadmap of what you plan to implement in the refactoring?

**Comment by evnbr** [comment]
I had a rough version working locally from a few months ago, just working around my day job to find the time to dig back in and finish. Hesitant to promise anything but I think it'll be over the line by beginning of january.

**Comment by danielepolencic** [comment]
I guess this is sort of related, but happy to open a new issue if it's not the case.
One common issue that I've found so far is this:

![image](https://user-images.githubusercontent.com/1130499/102677857-7fa58b00-41df-11eb-987f-eefe2acc37ea.png)

The HTML of that content is roughly equal to:

```html
<div class="code-block">
  <nav>..</nav>
  <pre><code>
     <div> code line </div>
     ... one div per line ...
  </code></pre>
</div>
```

Ideally, I'd like the header to split only when it has at least 2 lines of code in the same page. Otherwise, it should continue in the next page.
I had a look at the code, and (at the moment) the `shouldtraverse` function works on one element at the time, whereas I think I need a way to decide to move to the next page based on surrounding elements too.

```ts
if (this.element has at least 2 lines of code && isOverflow()) {
  move everything to the new page.
}
```

I tried to come up with a sensible option on how to change the code to do that, and it's not trivial.
I think [this code](https://github.com/evnbr/regionize/blob/0a9cab68d00f0a505b205af1309d9e8314af80ff/src/tryInNextRegion.ts#L38) could be expanded to include a custom function that could decide if the element can be moved to the next region (but only after traversing and splitting).
The function could have the following signature:

```ts
function customCanSplit(parentElement: HTMLElement, proposedMoveableElement: HTMLElement): boolean {
 // here I can decide if proposedMoveableElement is worth splitting.
}
```

I don't think this is an easy change so I wanted to check with you first.

**Comment by evnbr** [comment]
I think what you're describing is similar to one of the earliest unresolved request with bindery (https://github.com/evnbr/bindery/issues/8), which is about something like [indesign "keep" options](http://ikkydesign.com/how-to-keep-text-together-when-separated-over-columns-in-indesign/).

the area of code you linked to is very fragile and was supposed to be a band-aid. basically, to achieve the effect you're going for, we need to backtrack the traversal process entirely. Among other problems, backtracking should means undoing any changes from rules that have been triggered... but right now rules can have arbitrary imperative side effects that aren't tracked. the goal of my refactor is to make rules much more predictable so this sort of thing is possible, but even then will not be easy.
